src/c/uCLib_base

```diff
@@ -0,0 +1,127 @@
+ *           Please note: This library is NOT a replacement for the
+ *           Acorn/ROOL  SharedCLib (SCL)  or  the  ANSILib (ANSL).
+ *           It is  intended for use  with applications that do not
+ *           need the  entire  SCL or ANSL,  that  wish to  have no
+ *           dependecies  on  the  SCL/ANSL and still use C for the
+ *           source coding.
+ *
+ *   author: Paolo Fabio Zaino
+ *  release: 0.0.0
+ * 
+ * Copyright 2021 by Paolo Fabio Zaino
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * Redistributions  of  source  code  must retain the above copyright
+ * notice,  this list of conditions, the distribution license and the
+ * following disclaimer.
+ * Redistributions   in   binary   form  must   reproduce  the  above
+ * copyright  notice,  this  list  of  conditions  and  the following
+ * disclaimer  in  the  documentation and/or other materials provided
+ * with the distribution.
+ * Neither  the  name of the author nor the names of the contributors
+ * may  be  used  to  endorse  or  promote products derived from this
+ * software without specific prior written permission.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License
+ * the restrictions above and the disclaimer below.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * THIS  SOFTWARE   IS  PROVIDED  BY   THE  COPYRIGHT   HOLDERS  AND
+ * CONTRIBUTORS  "AS IS"  AND  ANY  EXPRESS  OR  IMPLIED WARRANTIES,
+ * INCLUDING,  BUT  NOT  LIMITED  TO,   THE  IMPLIED  WARRANTIES  OF
+ * MERCHANTABILITY  AND  FITNESS  FOR  A   PARTICULAR   PURPOSE  ARE
+ * DISCLAIMED.
+ * IN NO EVENT SHALL THE  COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR  ANY  DIRECT,  INDIRECT,  INCIDENTAL,  SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT
+ * OF  SUBSTITUTE  GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
+ * OR  BUSINESS  INTERRUPTION)  HOWEVER  CAUSED AND ON ANY THEORY OF
+ * LIABILITY,  WHETHER  IN  CONTRACT,   STRICT  LIABILITY,  OR  TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
+ * USE OF  THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+ * DAMAGE.
+ *
+ */
+
+#include "uCLib.h"
+
+void sys_exit(int error_code)
+{
+	// TODO: Improve code to handle error block
+	__asm
+	{
+		MOV r0, #0
+		MOV r1, #ABEX
+		MOV r2, error_code
+		SWI OS_Exit, {r0-r2}, {}, {}
+	}
+}
+
+int sys_write_stdout(const char *buf, unsigned count)
+{
+	int rval = 0;
+	if ( count == 0 ) {
+		// Display a null terminated string
+		__asm
+		{
+			MOV r0, buf
+			SWI OS_Write0, {r0}
+		}
+	} else {
+		// Display a string of "count" lenght
+		__asm
+		{
+			STMFD sp, {r0,r1,r2}
+			MOV   a1, #0
+			MOV   a2, buf
+			MOV   a3, count
+		loop:
+			LDRB  a1, [a2], #1
```

Comment:
This will dereference the pointer buf when count == 0.

Reply:
Not sure I fully understand what you mean here, when count is == 0  we do not use this code. So are you saying that this would happens when count gets down to 0? well tested that and it seems to be ok, did you check the test?

Reply:
I meant that the LDRB will be executed count + 1 times. I missed that this won't happen if the function is called with count == 0.

Reply:
I see what you meant, thx for clarifying. So question are we absolutely sure that DDE __asm is translating that ```MOV a2, buf``` to exactly that? I am not sure to be honest. In the CUtilHellowWorld test you can see that I explicitly re-print the string pointed by the pointer and it seems to be ok after the first time it's being printed, that made me think that the pointer address is copied and not loaded by the same location. 

I'll run more tests, but right now it seems that the ASM in __asm is not translated one-o-one, if this makes sense.

Reply:
In this case, looking at the object code, that has been assembled as `MOV a2, a1` as to be expected. The `MOV a1, #0` has been ommitted, presumably because it saw it will always be overwritten by the LDRB. 

The post-increment has been ommitted - unexpectedly since that would make it not do what you told it to. Are you sure the object files were compiled from the source you submitted? In your commit you have a post-increment _and_ an ADD, causing it to skip characters, but the object file doesn't have the post-increment.

Sorry I don't follow what you mean about pointers.

Reply:
I must have forgotten the ADD, let me check, again the sys_* code it's just for testing purposes so, I did not spend much time on it (however I ma pretty sure I have never seen it skipping characters tbh). But ok I'll double check, the pushed version is the same I have, given that the temporary code has not been touched since it has started working.

Reply:
> Sorry I don't follow what you mean about pointers.

The pointer to the message buffer that should be de-referenced doesn't appears to be so when I re-use it again (without changing) to print the same message twice, so there seems to be a copy somewhere. I'll add to the original test 00 printing of the pointer address before and after the first print, just in case. the \0 at the end of the message is where the (count == 0)-1 should be at, so this is what should be happening right now, I'll check it later when I'll be in front of the code, right now I am working so total different machines in front of me, sorry.

Reply:
Ok this is interesting, I had to check it, so got a coffee break, fired up my panda and commented out the ADD, according to your description this time it should have generated the code with the post increment so it should have worked, BUT the function only printed HHHHHHHHHHH, so I am thinking that the __asm in-line may actually just be ignoring the post increment (regardless the ADD) and that's why it worked from the beginning. This is just an assumption, but it seems that it ignored the post-increment regardless, I will try more stuff later, just wanted to let you know this finding.

Reply:
Compiling does actually give a couple of warnings about register corruption. I took a look at the manual and it says you can't use the stack, so I'd strike the STMFD/LDMFD. I think we should also, as recommended, leave it to do the register choices as much as possible, so:
````
__asm
{
loop:
  LDRB  a1, [buf], #1
  SWI   OS_WriteC, {a1}
  SUBS  count, count, #1
  BNE   loop
}
````
which generates the right code, but I didn't figure out exactly what made the corruption before.

Reply:
already done that and opted for OS_WriteN actually so 3 lines of code, will push the changes later :)

I guess what was confusing the compiler was the use of the stack, in another function on next branch the post-increment is working fine and the only difference between the two is I am not trying to use the stack.